Gems/GradientSignal/Code/Source/Editor/EditorImageGradientComponentMode.cpp

```diff
@@ -27,9 +27,198 @@
+        // Given a pixel index, get the relative pixel index within the tile
+        uint32_t GetPixelTileIndex(const PixelIndex& pixelIndex) const
+        {
+            uint32_t xIndex = pixelIndex.first % ImageTileSize;
+            uint32_t yIndex = pixelIndex.second % ImageTileSize;
+
+            return (yIndex * ImageTileSize) + xIndex;
+        }
+
+        //! Create an image tile initialized with the image gradient values if it doesn't already exist.
+        void CreateImageTile(uint32_t tileIndex)
+        {
+            // If it already exists, there's nothing more to do.
+            if (m_paintedImageTiles[tileIndex])
+            {
+                return;
+            }
+
+            auto imageTile = AZStd::make_unique<ImageTile>();
+
+            // Initialize the list of pixel indices for this tile.
+            AZStd::array<PixelIndex, ImageTileSize * ImageTileSize> pixelIndices;
+            PixelIndex startIndex = GetStartPixelIndex(tileIndex);
+            for (int16_t index = 0; index < (ImageTileSize * ImageTileSize); index++)
+            {
+                pixelIndices[index] = PixelIndex(startIndex.first + (index % ImageTileSize), startIndex.second + (index / ImageTileSize));
+            }
+
+            AZ_Assert(imageTile->first.size() == pixelIndices.size(), "ImageTile and PixelIndices are out of sync.");
+
+            auto& [originalValues, modifiedValues] = (*imageTile);
+
+            // Read all of the original gradient values into the image tile buffer.
+            ImageGradientModificationBus::Event(
+                m_imageGradientEntityId, &ImageGradientModificationBus::Events::GetPixelValuesByPixelIndex, pixelIndices, originalValues);
+
+            // Initialize the modified value buffer with the original values. This way we can always undo/redo an entire tile at a time
+            // without tracking which pixels in the tile have been modified.
+            modifiedValues = originalValues;
+
+            m_paintedImageTiles[tileIndex] = AZStd::move(imageTile);
+
+            // If we create a tile, we'll use that as shorthand for tracking that changed data exists.
+            m_modifiedAnyPixels = true;
+        }
+
+        //! Size of each modified image tile that we'll cache off.
+        //! This size is chosen somewhat arbitrarily to keep the number of tiles balanced at a reasonable size.
+        static inline constexpr uint32_t ImageTileSize = 32;
+
+        //! Keeps track of all the unmodified and modified gradient values for an NxN tile.
+        //! We store it a pair of arrays instead of an array of pairs for better compatibility with the ImageGradient APIs,
+        //! where we can just pass in a full array of values to update a full tile of values at once.
+        using ImageTile = AZStd::pair<
+                AZStd::array<float, ImageTileSize * ImageTileSize>,         // Unmodified data
+                AZStd::array<float, ImageTileSize * ImageTileSize>>;        // Modified data
```

Comment:
nit: I think it'd be a little cleaner to just have a simple struct for this. Having to reference `first` / `second` isn't as straightforward as members named something like `m_originalData` / `m_modifiedData`;

Reply:
i'll replace with a struct in my next PR.